api/python/t4/packages.py

```diff
@@ -155,21 +156,80 @@ def _clone(self):
         return PackageEntry(copy.deepcopy(self.physical_keys), self.size, \
                             copy.deepcopy(self.hash), copy.deepcopy(self.meta))
 
+
+def parse_s3_url(s3_url):
+    parsed = urlparse(s3_url)
+    no_scheme = parsed.netloc + parsed.path
+    return no_scheme
+
 class Package(object):
     """ In-memory representation of a package """
 
-    def __init__(self, data=None, meta=None):
-        """
-        _data is of the form {logical_key: PackageEntry}
+    def __init__(self, name=None, pkg_hash=None, registry=None):
         """
-        self._data = data or {}
-        self._meta = meta or {}
+        Create a Package from scratch, or load one from a registry.
+
+        Args:
+            name(string): name of package to load
+            pkg_hash(string): top hash of package version to load
+            registry(string): location of registry to load package from
+        """
+        if name is None and pkg_hash is None and registry is None:
+            self._data = {}
+            self._meta = {'version': 'v0'}
+            return
+
+        if registry is None:
+            # default to local registry
+            registry = get_package_registry()
+
+        if pkg_hash is not None:
+            # if hash is specified, name doesn't matter
+            pkg_path = '{}/packages/{}'.format(registry, pkg_hash)
+            # TODO replace open with something that supports both local and s3
+            self = self._from_path(pkg_path)
+            return
+
+        pkg_path = '{}/named_packages/{}/'.format(registry, quote(name))
+        latest = pkg_path + 'latest'
+        if latest.startswith('file:///'):
```

Comment:
I just merged my PR with some URI code. Can you update this to parse the URL similar to `copy_file` in `data_transfer.py`? Specifically, use `url2pathname` and `parse_s3_url`.

I'm still trying to come up with some abstraction to avoid `if``/`elif` everywhere we deal with S3 and files... Not sure how to do it.

Reply:
As far as the if/elif chains go, I think factoring out interfaces like `StorageProvider`, etc will help us use dynamic dispatch instead of if/elif